Add guide to type narrowing (#1798)
python/typing: docs/source/type_narrowing.rst

**************
Type Narrowing
**************

Python programs often contain symbols that take on multiple types within a
single given scope and that are distinguished by a conditional check at
runtime. For example, here the variable *name* can be either a ``str`` or
``None``, and the ``if name is not None`` narrows it down to just ``str``::

    def maybe_greet(name: str | None) -> None:
        if name is not None:
            print("Hello, " + name)

This technique is called *type narrowing*.
To avoid false positives on such code, type checkers understand
various kinds of conditional checks that are used to narrow types in Python code.
The exact set of type narrowing constructs that a type checker understands

> is not specified and varies across type checkers. Commonly understood

Reviewer: Should we have a minimum required set of narrowing constructs? The one bellow might as well be one with a few additions, e.g. `issubclass`.

Reviewer 2: > Should we have a minimum required set of narrowing constructs?

That would require a change to the spec, which would have to go through the typing council. This PR doesn't propose a change to the spec, it just adds a new page of user-facing documentation (it's descriptive, not prescriptive).

Reviewer: Sure, I guess this is a request for a spec update then :)

Author: Covering this behavior in the spec is probably useful but it's a much larger project. This PR is meant as user-facing documentation, not a full spec.

patterns include:

* ``if x is not None``
* ``if x``

> * ``if isinstance(x, SomeType)``

Reviewer: Would it be useful to have an example of narrowing to a subclass? E.g.

```
def f(x: Base):
  if isinstance(x, SomeImpl): ...
```

Author: Since this behavior is unspecified and `isinstance()` behavior does vary across type checkers, I'd prefer not to go into too much detail.

* ``if callable(x)``

In addition to narrowing local variables, type checkers usually also support
narrowing instance attributes and sequence members, such as

> ``if x.some_attribute is not None`` or ``if x[0] is not None``, though the exact

Reviewer: Is it worth saying that such narrowing is potentially unsound but still useful in practice?

Author: As the document explains later, virtually all narrowing is potentially unsound.

Reviewer: It might still be useful to explain where the unsoundness comes from. E.g. I think something like `if x is not None` can only be unsound in the presence of dynamic features, such as `globals()["x"] = ...`.

Reviewer: Thanks, I somehow missed it during my initial review. The section looks good.

conditions for this behavior differ between type checkers.

Consult your type checker's documentation for more information on the type
narrowing constructs it supports.

The type system also includes two ways to create *user-defined* type narrowing
functions: :py:data:`typing.TypeIs` and :py:data:`typing.TypeGuard`. These
are useful if you want to reuse a more complicated check in multiple places, or
you use a check that the type checker doesn't understand. In these cases, you
can define a ``TypeIs`` or ``TypeGuard`` function to perform the check and allow type checkers

> to use it to narrow the type of a variable. Among the two, ``TypeIs`` usually

Reviewer 3: nit: for only two items, I would usually prefer "Between" over "Among"
```suggestion
to use it to narrow the type of a variable. Between the two, ``TypeIs`` usually
```

* ``TypeGuard`` is in ``typing`` starting from Python 3.10 and in ``typing_extensions``
  starting from version 3.10.0.0.

Writing a correct ``TypeIs`` function
-------------------------------------

A ``TypeIs`` function allows you to override your type checker's type narrowing
behavior. This is a powerful tool, but it can be dangerous because an incorrectly
written ``TypeIs`` function can lead to unsound type checking, and type checkers
cannot detect such errors.

For a function returning ``TypeIs[T]`` to be correct, it must return ``True`` if and only if
the argument is compatible with type ``T``, and ``False`` otherwise. If this condition is
not met, the type checker may infer incorrect types.

Below are some examples of correct and incorrect ``TypeIs`` functions::

    from typing import TypeIs

    # Correct
    def is_int(x: object) -> TypeIs[int]:
        return isinstance(x, int)

    # Incorrect: does not return True for all ints
    def is_positive_int(x: object) -> TypeIs[int]:
        return isinstance(x, int) and x > 0

    # Incorrect: returns True for some non-ints
    def is_real_number(x: object) -> TypeIs[int]:
        return isinstance(x, (int, float))

> 

Reviewer 4: These two examples would also be good to have in the `TypeIs` documentation at docs.python.org. Also, the `TypeIs` and `TypeGuard` docs could link to this document when it gets merged.

            isinstance(obj, dict)
            and all(isinstance(key, str) for key in obj)
            and isinstance(obj.get("x"), int)
            and isinstance(obj.get("y"), int)
        )

.. _`guide-type-narrowing-typeis-typeguard`:

``TypeIs`` and ``TypeGuard``
----------------------------

:py:data:`typing.TypeIs` and :py:data:`typing.TypeGuard` are both tools for narrowing the type of a variable
based on a user-defined function. Both can be used to annotate functions that take an
argument and return a boolean depending on whether the input argument is compatible with
the narrowed type. These function can then be used in ``if`` checks to narrow the type
of a variable.

``TypeIs`` usually has the more intuitive behavior, but it
introduces more restrictions. ``TypeGuard`` is the right tool to use if:

* You want to narrow to a type that is not compatible with the input type, for example
  from ``list[object]`` to ``list[int]``.  ``TypeIs`` only allows narrowing between
  compatible types.
* Your function does not return ``True`` for all input values that are compatible with
  the narrowed type. For example, you could have a ``TypeGuard[int]`` that returns ``True``

>   only for positive integers.

Reviewer 3: Since you already use "assignable to" below (and link to the glossary), maybe prefer "assignable to" over "compatible with" in these points as well?

Author: Thanks, good point. I wrote that section of the PEP before your spec change :)

Author: There is also a usage of "compatible" on line 87, but there it refers to runtime compatibility so I don't think "assignable" is right; there I'll simply say "is of type T".

**************
Type Narrowing
**************

> Python programs often contain values that can take on multiple types and
> that are distinguished by a conditional check at runtime. For example, here
> the variable *name* can be either a ``str`` or ``None``, and the
> ``if name is not None`` narrows it down to just ``str``::

Reviewer 2: ```suggestion
Python programs often contain symbols that take on multiple types within a
single given scope and that are distinguished by a conditional check at
runtime. For example, here the variable *name* can be either a ``str`` or
``None``, and the ``if name is not None`` narrows it down to just ``str``::
```

    def maybe_greet(name: str | None) -> None:
        if name is not None:
            print("Hello, " + name)

This technique is called *type narrowing*.
In order to avoid false positives on such code, type checkers understand
various kinds of conditionals that are used to narrow types in Python code.
The exact set of type narrowing constructs that a type checker understands

> is not specified and varies across type checkers. Commonly understood type

Author: I went with just "patterns"; that seems clear in context.

patterns include:

* ``if x is not None``
* ``if x``
* ``if isinstance(x, SomeType)``
* ``if callable(x)``

In addition to narrowing local variables, type checkers usually also support
narrowing instance attributes and sequence members, such as

> ``if x.some_attribute is not None`` or ``if x[0] is not None``.

Author: Pyre seems to do type narrowing here too: https://pyre-check.org/play?input=%23%20pyre-strict%0A%0Afrom%20typing%20import%20*%0A%0Afrom%20dataclasses%20import%20dataclass%0A%0A%40dataclass%0Aclass%20X%3A%0A%20%20%20%20a%3A%20int%20%7C%20None%0A%0Adef%20f(x%3A%20X)%20-%3E%20None%3A%0A%20%20%20%20if%20x.a%20is%20not%20None%3A%0A%20%20%20%20%20%20%20%20reveal_type(x.a)%0A%20%20%20%20%20%20%20%20print(x.a%20%2B%201)%0A%20%20%20%20print(x.a%20%2B%201)

I'm not sure why this shouldn't be mentioned. It's something users commonly need and ask for. While it is somewhat more obviously unsafe than some other narrowing patterns, there's few kinds of type narrowing that are 100% safe.

> ``if x.some_attribute is not None`` or ``if x[0] is not None``.

Reviewer 2: > Pyre seems to do type narrowing here too:

Correct, but if you insert a single intermediate function call, it decides that it can no longer assume that the attribute wasn't modified in between the two accesses by the intermediate function call, and no longer does the type narrowing: https://pyre-check.org/play?input=%23%20pyre-strict%0A%0Afrom%20typing%20import%20*%0A%0Afrom%20dataclasses%20import%20dataclass%0A%0A%40dataclass%0Aclass%20X%3A%0A%20%20%20%20a%3A%20int%20%7C%20None%0A%0Adef%20do_something_else()%20-%3E%20None%3A%0A%20%20%20%20pass%0A%0Adef%20f(x%3A%20X)%20-%3E%20None%3A%0A%20%20%20%20if%20x.a%20is%20not%20None%3A%0A%20%20%20%20%20%20%20%20do_something_else()%0A%20%20%20%20%20%20%20%20reveal_type(x.a)%0A%20%20%20%20%20%20%20%20print(x.a%20%2B%201)%0A%20%20%20%20print(x.a%20%2B%201)

> I'm not sure why this shouldn't be mentioned.

Well, I just worry about documenting behaviour here that isn't standardised between all major type checkers. It might be pretty confusing for users of type checkers with differing behaviour.

> ``if x.some_attribute is not None`` or ``if x[0] is not None``.

Author: The text is already pretty clear that the exact type narrowing behaviors differ among type checkers, but I'll add another line to clarify that.

I think it's important to mention this behavior because it's a common pattern in practice, so people are likely to run into this.

> ``if x.some_attribute is not None`` or ``if x[0] is not None``.

Reviewer 2: Thanks, I think the extra line definitely helps

    def bad_typeis1(x: object) -> TypeIs[int]:
        return isinstance(x, int) and x > 0

    # Incorrect: returns True for some non-ints
    def bad_typeis2(x: object) -> TypeIs[int]:
        return isinstance(x, (int, float))

This function demonstrates some errors that can occur when using a poorly written
``TypeIs`` function. These errors are not detected by type checkers::

    def caller(x: int | str, y: int | float) -> None:
        if bad_typeis1(x):  # narrowed to int
            print(x + 1)
        else:  # narrowed to str (incorrectly)
            print("Hello " + x)  # runtime error if x is a negative int

        if bad_typeis2(y):  # narrowed to int
            # Because of the incorrect TypeIs, this branch is taken at runtime if
            # y is a float.
            print(y.bit_count())  # runtime error: this method exists only on int, not float
        else:  # narrowed to float (though never executed at runtime)
            pass

Here is an example of a correct ``TypeIs`` function for a more complicated type::

    from typing import TypedDict, TypeIs

    class Point(TypedDict):
        x: int
        y: int

>     def is_point(x: object) -> TypeIs[Point]:
>         return (
>             isinstance(x, dict)
>             and all(isinstance(key, str) for key in x)
>             and "x" in x
>             and "y" in x
>             and isinstance(x["x"], int)
>             and isinstance(x["y"], int)
>         )

Reviewer 2: ```suggestion
    def is_point(obj: object) -> TypeIs[Point]:
        return (
            isinstance(obj, dict)
            and all(isinstance(key, str) for key in obj)
            and isinstance(obj.get("x"), int)
            and isinstance(obj.get("y"), int)
        )
```

.. _`guide-type-narrowing-typeis-typeguard`:

``TypeIs`` and ``TypeGuard``
----------------------------

:py:data:`typing.TypeIs` and :py:data:`typing.TypeGuard` are both tools for narrowing the type of a variable
based on a user-defined function. Both can be used to annotate functions that take an
argument and return a boolean depending on whether the input argument is compatible with
the narrowed type. These function can then be used in ``if`` checks to narrow the type
of a variable.

> ``TypeIs`` usually has the most intuitive behavior, but it

Reviewer 2: ```suggestion
``TypeIs`` usually has the more intuitive behavior, but it
```

introduces more restrictions. ``TypeGuard`` is the right tool to use if:

* You want to narrow to a type that is not compatible with the input type, for example
  from ``list[object]`` to ``list[int]``.  ``TypeIs`` only allows narrowing between
  compatible types.
* Your function does not return ``True`` for all input values that are compatible with
  the narrowed type. For example, you could have a ``TypeGuard[int]`` that returns ``True``
  only for positive integers.

``TypeIs`` and ``TypeGuard`` differ in the following ways:

* ``TypeIs`` requires the narrowed type to be a subtype of the input type, while
  ``TypeGuard`` does not.
* When a ``TypeGuard`` function returns ``True``, type checkers narrow the type of the
  variable to exactly the ``TypeGuard`` type. When a ``TypeIs`` function returns ``True``,
  type checkers can infer a more precise type combining the previously known type of the

>   variable with the ``TypeIs`` type. (Technically, this is known as an intersection type.)

Reviewer 2: Not sure "technically" adds much here: the whole document is technical :)

```suggestion
  variable with the ``TypeIs`` type. (This is known as an "intersection type".)
```

You could possibly say

```suggestion
  variable with the ``TypeIs`` type. (In type theory, this is known as an
  "intersection type".)
```

Author: Thanks, I think the first suggestion is best (when in doubt, I like to go with the less verbose version).

at runtime.::

    from typing import TypeIs

    def is_str(x: object) -> TypeIs[str]:
        return True

    def takes_str_or_int(x: str | int) -> None:
        if is_str(x):
            print(x + " is a string")  # runtime error

To avoid this problem, every ``TypeIs`` and ``TypeGuard`` function should be
carefully reviewed and tested.

Unsound ``TypeGuard`` narrowing
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Unlike ``TypeIs``, ``TypeGuard`` can narrow to a type that is not a subtype of the
original type. This allows for unsafe behavior with invariant data structures::

    from typing import Any, TypeGuard

    def is_int_list(x: list[Any]) -> TypeGuard[list[int]]:
        return all(isinstance(i, int) for i in x)

    def maybe_mutate_list(x: list[Any]) -> None:
        if is_int_list(x):
            x.append(0)  # OK, x is narrowed to list[int]

    def takes_bool_list(x: list[bool]) -> None:
        maybe_mutate_list(x)
        reveal_type(x)  # list[bool]
        assert all(isinstance(i, bool) for i in x)  # fails at runtime

    takes_bool_list([True, False])

To avoid this problem, use ``TypeIs`` instead of ``TypeGuard`` where possible.

> If you must use ``TypeGuard``, avoid narrowing across incompatible types.

Reviewer 2: Possibly some better (and certainly more general) advice is to try (though it's not always possible) to use covariant, immutable types for parameter annotations wherever possible. If you do so, there will be many more cases where you'll be able to use `TypeIs`, and so the problem here just arises much less in the first place

Author: Maybe this?
```suggestion
To avoid this problem, use ``TypeIs`` instead of ``TypeGuard`` where possible.
If you must use ``TypeGuard``, avoid narrowing across incompatible types.
Prefer using covariant, immutable types in parameter annotations (e.g.,
``Sequence`` or ``Iterable`` instead of ``list``. If you do this, it is more likely
that you'll be able to use ``TypeIs`` to implement your type narrowing functions.
```

Reviewer 2: That looks good to me! Though you need a closing parenthesis after "instead of `list`"

Reviewer 2: It looks like you marked this as resolved but didn't push the change, was that deliberate? :)
